### PR TITLE
refactor(metric): Speedup AUPRO

### PR DESCRIPTION
**Pull request by alfieroddanintel**
## 📝 Description

This PR introduces a new, significantly faster implementation of AUPRO inspired by the [evaluation code of MVTecAD](https://www.mvtec.com/company/research/datasets/mvtec-ad).

## ✨ Changes

Instead of computing a ROC curve for every individual connected component (the previous approach), the new implementation performs a single global sort of anomaly scores and derives both FPR and per-region TPR contributions in one pass.

This removes thousands of per-region indexing operations, eliminates repeated calls to binary_roc, and avoids expensive boolean masking and slicing inside Python loops.

## Old vs New

To compare the two implementations I take the worst case scenario random predictions and ground truth. This results in lots and lots of computation.

## CPU Performance

| Resolution | Old Time (ms) | New Time (ms) | **Speedup** | Old AUPRO  | New AUPRO  |
|-----------|---------------:|---------------:|------------:|-----------:|-----------:|
| **128×128**   | 12501.30      | 129.68        | **96.4×**   | 0.147060   | 0.147062   |
| **256×256**   | 191969.17     | 872.97        | **220×**    | 0.149794   | 0.149784   |
| **512×512**   | —             | 13067.69      | —           | —          | 0.151356   |
| **1024×1024** | —             | —             | —           | —          | —          |

---

## CUDA Performance

| Resolution | Old Time (ms) | New Time (ms) | **Speedup** | Old AUPRO  | New AUPRO  |
|-----------|---------------:|---------------:|------------:|-----------:|-----------:|
| **128×128**   | 2924.40       | 412.90        | **7.1×**    | 0.160438   | 0.160441   |
| **256×256**   | 9031.21       | 124.13        | **72.7×**   | 0.151356   | 0.151345   |
| **512×512**   | 42281.08      | 476.24        | **88.8×**   | 0.149030   | 0.149071   |
| **1024×1024** | —             | 8453.04       | —           | —          | 0.151388   |

Feel free to check yourself.

```python
import time
import torch
from anomalib.data import ImageBatch
from anomalib.metrics.aupro import AUPRO

# DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
DEVICE = "cpu"


def fake_batch(bs=4, h=64, w=64, device=DEVICE):
    torch.manual_seed(0)
    return ImageBatch(
        image=torch.rand(bs, 3, h, w, device=device),
        anomaly_map=torch.rand(bs, 1, h, w, device=device),
        pred_score=torch.rand(bs, device=device),
        gt_label=(torch.rand(bs, device=device) > 0.5).int(),
        gt_mask=(torch.rand(bs, 1, h, w, device=device) > 0.97).long(),
    )


def run_aupro(metric_cls, batch):
    metric = metric_cls(fields=["anomaly_map", "gt_mask"], fpr_limit=0.3).to(DEVICE)

    t0 = time.time()
    metric.update(batch)
    value = metric.compute().item()
    elapsed_ms = (time.time() - t0) * 1000

    return value, elapsed_ms


def main():
    resolutions = [
        (128, 128),
        (256, 256),
        (512, 512),
        (1024, 1024),
    ]

    print("\n=== AUPRO Timing ===\n")
    print(f"{'Res':<12}{'Time (ms)':<12}{'AUPRO':<12}")

    for H, W in resolutions:
        batch = fake_batch(bs=4, h=H, w=W, device=DEVICE)

        value, ms = run_aupro(AUPRO, batch)

        print(f"{H}x{W:<8} {ms:>10.2f}   {value:0.6f}")


if __name__ == "__main__":
    torch.manual_seed(0)
    main()

```

Select what type of change your PR is:

- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔄 Refactor (non-breaking change which refactors the code base)
- [ ] ⚡ Performance improvements
- [ ] 🎨 Style changes (code style/formatting)
- [ ] 🧪 Tests (adding/modifying tests)
- [ ] 📚 Documentation update
- [ ] 📦 Build system changes
- [ ] 🚧 CI/CD configuration
- [ ] 🔧 Chore (general maintenance)
- [ ] 🔒 Security update
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [x] 📚 I have made the necessary updates to the documentation (if applicable).
- [x] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).
- [x] 🏷️ My PR title follows conventional commit format.

For more information about code review checklists, see the [Code Review Checklist](https://github.com/open-edge-platform/anomalib/blob/main/docs/source/markdown/guides/developer/contributing.md).

### CODE REVIEW

**Comment by rajeshgangireddy** [comment]
Wow. This is great. 
Just one question, any hypothesis on why computation of AUPRO is slower on cuda at 128x128 than that on CPU  ( 412.90 vs 129.68 )?

**Comment by alfieroddanintel** [comment]
> Wow. This is great. Just one question, any hypothesis on why computation of AUPRO is slower on cuda at 128x128 than that on CPU ( 412.90 vs 129.68 )?

Nothing for certain but probably a mix of random data is not that easy for GPU and also the benchmark script being very precise, more like trends.
